File: guys-long-connection/src/main/java/com/imuguys/im/connection/SocketMessageListenerGroup.java
package com.imuguys.im.connection;

import java.util.HashSet;

/**
 * 监听相同Message类型的SocketMessageListener组
 */
public class SocketMessageListenerGroup<Message> implements SocketMessageListener<Message> {

  public SocketMessageListenerGroup(String messageClassName) {
    mMessageClassName = messageClassName;
  }

  private String mMessageClassName;
  private HashSet<SocketMessageListener<Message>> mSocketMessageListenerSet = new HashSet<>();

  void addMessageListener(SocketMessageListener<Message> messageListener) {
    mSocketMessageListenerSet.add(messageListener);
  }

  void removeMessageListener(SocketMessageListener<Message> messageListener) {
    mSocketMessageListenerSet.remove(messageListener);
  }

  @Override
  public void handleMessage(Message message) {
    mSocketMessageListenerSet.forEach(socketMessageListener -> {
      socketMessageListener.handleMessage(message);
    });
  }

  public String getMessageClassName() {
    return mMessageClassName;
  }
}

File: guys-long-connection/src/main/java/com/imuguys/im/connection/SocketMessageListenerGroup.kt
package com.imuguys.im.connection

/**
 * 监听相同Message类型的SocketMessageListener组
 */
class SocketMessageListenerGroup<Message>(messageClassName: String) :
    SocketMessageListener<Message> {
    override fun handleMessage(message: Message) {
        mSocketMessageListenerSet.forEach {
            it.handleMessage(message)
        }
    }

    val mMessageClassName: String = messageClassName
    var mSocketMessageListenerSet = HashSet<SocketMessageListener<Message>>()

    fun addMessageListener(messageListener: SocketMessageListener<Message>) {
        mSocketMessageListenerSet.add(messageListener)
    }

    fun removeMessageListener(messageListener: SocketMessageListener<Message>) {
        mSocketMessageListenerSet.remove(messageListener)
    }
}
